RFC: when loading code for internal purposes, load stdlib files directly, bypassing DEPOT_PATH, LOAD_PATH, and stale checks

vtjnash opened the pull request:
This helps avoid the situation that the user can break important Pkg, REPL, Socket or similar features simply because they chose to remove `@stdlib` from their environment. For example, if you make an edit to REPL, then this will trigger recompilation and load the edited version:
```
$ ./julia -e 'using REPL' -iq
┌ Info: Precompiling REPL [3fa0cd96-eef1-5676-8a61-b3b8758bbffb] (cache misses: include_dependency fsize change (2), invalid header (10), mismatched flags (1))
└ @ Base loading.jl:2643
julia>
```
But this will load the version that shipped with Julia, regardless of the state of the cache or the source code (unless you delete the cache files):
```
$ ./julia -iq
julia>
```

Fixes #53365

vtjnash committed:
for internal purposes, load stdlib files directly, bypassing DEPOT_PATH, LOAD_PATH, and stale checks

Helps avoid the situation that the user can break important Pkg, REPL,
Socket or similar features simply by choosing to remove @stdlib from
their environment.

vtjnash commented:
For an added example, see here where I set everything to be invalid, yet Julia doesn't break:
```
$ JULIA_DEPOT_PATH=. JULIA_PROJECT=. JULIA_LOAD_PATH=. ./julia 
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.11.0-DEV.1591 (2024-02-13)
 _/ |\__'_|_|_|\__'_|  |  master/74349d33641* (fork: 1 commits, 0 days)
|__/                   |

julia> using REPL
ERROR: ArgumentError: Package REPL not found in current path.
- Run `import Pkg; Pkg.add("REPL")` to install the REPL package.
Stacktrace:
 [1] macro expansion
   @ ./loading.jl:1908 [inlined]
 [2] macro expansion
   @ ./lock.jl:273 [inlined]
 [3] __require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1883
 [4] #invoke_in_world#3
   @ ./essentials.jl:1064 [inlined]
 [5] invoke_in_world
   @ ./essentials.jl:1061 [inlined]
 [6] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1876

julia> Base.LOAD_PATH
1-element Vector{String}:
 "."

julia> Base.DEPOT_PATH
1-element Vector{String}:
 "."

julia> Base.active_project()
"/home/vtjnash/julia/Project.toml"
```

But note that `Pkg.REPLExt` is still broken, as I inserted a call to the wrong `insert_extension_triggers` function, and using `locate_package_env` as an example is not helpful to derive the correct argument, as that function currently returns the explicitly _wrong_ value for this particular case

KristofferC commented:
> But note that Pkg.REPLExt is still broken, as I inserted a call to the wrong insert_extension_triggers function, and using locate_package_env as an example is not helpful to derive the correct argument, as that function currently returns the explicitly wrong value for this particular case

Could you elaborate on this with an example perhaps? I don't fully understand what the actual issue is.

vtjnash committed:
some NFCI loading rearrangements

vtjnash committed:
defer callbacks after precompile loading

Preferable for all precompile code to finish loading, and not interrupt
it to keep running various callbacks.
